examples/wave/wave-op-mpi.py

```diff
@@ -192,14 +196,16 @@ def main(ctx_factory, dim=2, order=3, visualize=False):
     def rhs(t, w):
         return wave_operator(dcoll, c=c, w=w)
 
+    compiled_rhs = actx.compile(rhs)
+
     if comm.rank == 0:
         logger.info("dt = %g", dt)
 
     t = 0
     t_final = 3
     istep = 0
     while t < t_final:
-        fields = rk4_step(fields, t, dt, rhs)
+        fields = rk4_step(fields, t, dt, compiled_rhs)
```

Comment:
I think we might need to still perform the `thaw/freeze` pattern as illustrated in meshmode's simple-dg example. Maybe @inducer has more to say on this?
```suggestion
        fields = thaw(freeze(fields, actx), actx)
        fields = rk4_step(fields, t, dt, compiled_rhs)
```

Actually, I think my suggestion might not be the optimal way. The right way is probably illustrated here: https://github.com/inducer/meshmode/blob/main/examples/simple-dg.py#L494-L496 --- meaning we just wrap the rhs evaluations in a thaw/freeze pattern.

Reply:
Hmm, but isn't the thaw/freeze pattern due to using two different actx (like simple-dg does? 

I'm also confused why the ESDG example just thaws/freezes on the same actx:
https://github.com/inducer/grudge/blob/49bcb38ba6de002eeec8fb383998877339e7a753/examples/euler/acoustic_pulse.py#L209-L210

Reply:
I was under the impression that this was necessary to keep pytato's compiler from doing some unnecessary/repeated code generation between timesteps (which is an incredibly frustrating issue itself). I recall this being an issue for `mirgecom` as well for @MTCam's temperature solver. Perhaps I am the one misunderstanding then. Would really appreciate @kaushikcfd or @inducer to enlighten us on this.

If I _don't_ do the thaw/freeze in between calls to the timestepper, I experienced repeated code-generation that kills the simulation runtime. That's the summary of my issue. And I would expect that this example likely also suffers from that. I haven't checked out this branch to run it (yet) but are you not experiencing this problem with the example as is?

Reply:
I was under the impression that this was necessary to keep pytato's compiler from doing some unnecessary/repeated code generation between timesteps (which is an incredibly frustrating issue itself). I recall this being an issue for `mirgecom` as well for the temperature solver. Perhaps I am the one misunderstanding then

If I _don't_ do the thaw/freeze in between calls to the timestepper, I experienced repeated code-generation that kills the simulation runtime. That's the summary of my issue. And I would expect that this example likely also suffers from that. I haven't checked out this branch to run it (yet) but are you not experiencing this problem with the example as is?

Reply:
I see, you're right! Should be fixed in 3e710a8

Reply:
For the PytatoPyOpenCLArrayContext, if we don't do this the fields starts becoming expression as below:
```
# at istep = 1
fields = fields(0) + dt/6*(k1(0) + 2*k2(0) + 2*k3(0)+k4(0))
# at istep = 2
fields = fields(0) + dt/6*(k1(0) + 2*k2(0) + 2*k3(0)+k4(0)) + dt/6*(k1(dt) + 2*k2(dt) + 2*k3(dt) + k4(dt))
```
I.e. the expression keeps growing in size. This is a consequence of us not thunking the values onto the expression objects. I agree this is a sub-optimal user-interface (Very related: https://github.com/inducer/arraycontext/issues/102).